trezor/trezor-suite
# feat(suite-native): pin handling

I recommend doing a review by commits not in whole 💡 


## Description

Better handling of unlocking the device. 
- Request PIN entry whenever Suite needs device and it's locked.
- Add PIN screen for all models, not only T1. 
- Allow to cancel PIN entry (close screen by X button).
- Handle 3rd wrong PIN attempt and request new PIN entry automatically for the user.
- Submitting state has been added in separate PR https://github.com/trezor/trezor-suite/pull/11198


TODOs:
- [x] Add missing animation for PIN submit button reveal.
- [x] ~Add missing alert when user cancel PIN entry during discovery.~ it breaks the discovery, same as on Desktop
- [x] Test lock device during discovery, I feel like it breaks the discovery.
- [x] Add screenshots and video.
- [x] BUG (acceptable?): with locked T1 in receive flow, there is no authFailed after 3rd wrong attempt so app navigate back and user has to click to reveal receive address again.
- [x] BUG (acceptable?): with T1 after PIN cancelation and then Connect & unlock, there is empty wallet state visible before the discovery starts.
- [x] BUG: if close button X is clicked after PIN submit, empty wallet state is visible before the discovery starts
- [x] Do not go to PIN entry screen before onboarding finished
- [x] If biometrics enabled, wait with PIN request until unlocked
- [ ] Go through all the testing scenarios again.

Notes for @trezor/qa:

<details>
  <summary> testing scenarios</summary>
  
T1,TT,TS3 - every scenario on every model, with and without biometrics

- onboarded app, connect trezor,
	- TT/TS3 enter PIN on device - suite see it only then
	- T1
		- [PIN ENTRY VARIANTS]
- app with trezor connected and unlocked, trezor gets locked (hodl screen TT, hold button TS3/T1), verify receive address
	- [PIN ENTRY VARIANTS]
- app with trezor connected and unlocked, trezor gets locked (short locktime) before discovery finishes device communication
	- [PIN ENTRY VARIANTS]
	- discovery should continue if valid PIN entered, stop if modal closed
- start app with trezor already connected previously
	- locked device before app start
	- unlocked device before app start
- start app with trezor unlocked
	- fresh app with onboarding
	- onboarded app
- connect trezor without PIN
	- fresh app with onboarding
	- onboarded app
- having some portfolio accounts, connect locked device, try cancel and valid PIN
		
		
[PIN ENTRY VARIANTS]
  - enter valid PIN
  - enter invalid PIN and then valid
  - enter invalid PIN more than 3 times and then valid
  - enter wipe code
  - close PIN entry modal
</details>
	

## Related Issue


Resolve https://github.com/trezor/trezor-suite/issues/10595 https://github.com/trezor/trezor-suite/issues/10596 https://github.com/trezor/trezor-suite/issues/10359 https://github.com/trezor/trezor-suite/issues/10536 https://github.com/trezor/trezor-suite/issues/10538

## Screenshots:
<img width="300" alt="image" src="https://github.com/trezor/trezor-suite/assets/3729633/371a5895-6ef4-41ec-9f07-44c998880370">
<img width="300" alt="image" src="https://github.com/trezor/trezor-suite/assets/3729633/a26fde88-4bb5-4b54-8b75-18d46682d4fa">


all PIN attempts wasted not reflected in Suite (there is no info about it from the device), but reflected on the device properly:
<img width="250" alt="image" src="https://github.com/trezor/trezor-suite/assets/3729633/e37f65e8-0bb7-4c97-bc54-6d251fa04601">
